### PR TITLE
Equiv hash map

### PR DESCRIPTION
**Pull request by FWDekker**
Implements the `CustomHashMap` and `CustomHashSet` classes, which are (respectively) `HashMap` and `HashSet` implementations that accept a custom hash method.

These classes will be used as described in my suggestion on #111.

### CODE REVIEW

File: modules/models/src/main/kotlin/org/cafejojo/schaapi/models/CustomHashCollections.kt [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    override fun remove(key: K) = innerMap.remove(wrapKey(key))
+
+    private fun wrapKey(key: K) = HashWrapper(key, customHash)
+
+    /**
+     * A simple implementation of [MutableMap.MutableEntry].
+     */
+    internal class Entry<K, V>(override val key: K, override var value: V) : MutableMap.MutableEntry<K, V> {
+        override fun setValue(newValue: V): V {
+            val oldValue = value
```
Comment (casperboone) [inline modules/models/src/main/kotlin/org/cafejojo/schaapi/models/CustomHashCollections.kt]:
```kotlin
=  value.also { value = newValue }
```

File: modules/models/src/test/kotlin/org/cafejojo/schaapi/models/CustomHashCollectionsTest.kt [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+        beforeEachTest {
+            map = CustomHashHashMap(Any::hashCode)
+        }
+
+        it("starts out empty") {
+            assertThat(map)
+                .hasSize(0)
+                .isEmpty()
+            assertThat(map.entries)
+                .isEmpty()
```
Comment (casperboone) [inline modules/models/src/test/kotlin/org/cafejojo/schaapi/models/CustomHashCollectionsTest.kt]: Can be on previous line

File: modules/models/src/test/kotlin/org/cafejojo/schaapi/models/CustomHashCollectionsTest.kt [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+class CustomHashMapTest : Spek({
+    describe("custom hash map") {
+        lateinit var map: CustomHashHashMap<Any, Any>
+
+        beforeEachTest {
+            map = CustomHashHashMap(Any::hashCode)
+        }
+
+        it("starts out empty") {
+            assertThat(map)
```
Comment (casperboone) [inline modules/models/src/test/kotlin/org/cafejojo/schaapi/models/CustomHashCollectionsTest.kt]: For me it's fine if it looks like
```kotlin
assertThat(map).hasSize(0).isEmpty()
```

File: modules/models/src/main/kotlin/org/cafejojo/schaapi/models/CustomHashCollections.kt [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    override fun clear() = innerMap.clear()
+
+    override fun put(key: K, value: V) = innerMap.put(wrapKey(key), value)
+
+    override fun putAll(from: Map<out K, V>) = innerMap.putAll(from.map { (key, value) -> Pair(wrapKey(key), value) })
+
+    override fun remove(key: K) = innerMap.remove(wrapKey(key))
+
+    private fun wrapKey(key: K) = HashWrapper(key, customHash)
```
Comment (gandreadis) [inline modules/models/src/main/kotlin/org/cafejojo/schaapi/models/CustomHashCollections.kt]: What fun!

**Pull request merged**